fix(api): Exclude existing module offsets when running subsequent module calibrations

vegano1 opened the pull request:
# Overview

Module Calibrations were failing because the calibration starting position was being shifted by the previous module calibration value for the given module, which meant that eventually the starting position would be way off-center and would not be able to find the calibration square. This PR addresses this issue and fixes another issue with non-connected modules (magnetic block) where we would throw `ModuleNotConnectedError` when trying to get the labware well position.

Closes: [RCORE-738](https://opentrons.atlassian.net/browse/RCORE-738)

# Test Plan

- [x] Make sure subsequent module calibrations start at the same position for each module type
- [x] Make sure we don't throw `ModuleNotConnectedError` when getting the position for labware on top of the magnetic block.

# Changelog

- Fixed an issue where the starting point for module calibrations would be off by the previous module offset causing subsequent module calibrations to fail.
- Fixed a potential issue in get_module_offset if the module is a non-connected module where just getting the well position would throw ModuleNotConnectedError, instead, we want to return the offset without the calibrated offset value.
- Added get_raw_module_offset so we can get the loaded module offsets without transforms
- Refactored the calibrate_module test script so it handles needing to home after robot-server restart, resetting states, and generally handling failure cases better.

# Review requests

Make sure we are comfortable with the get_module_changes which affect get_well_position

# Risk assessment

medium, while this affects module calibration directly any changes to get_module_offset also affect get_well_position


[RCORE-738]: https://opentrons.atlassian.net/browse/RCORE-738?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

vegano1 committed:
fix(api): Remove module offsets when running subsequent module calibrations

vegano1 committed:
fix unit test

sfoster1 commented on api/src/opentrons/protocol_engine/state/modules.py:
```diff
@@ -630,6 +630,15 @@ def get_dimensions(self, module_id: str) -> ModuleDimensions:
         """Get the specified module's dimensions."""
         return self.get_definition(module_id).dimensions
 
+    def get_raw_module_offset(self, module_id: str) -> ModuleOffsetVector:
```
if we're modifying PE for this, I think it's probably more straightforward to add like `get_nominal_well_position` and `get_nominal_module_position`.

sanni-t replied:
I like that!

sfoster1 commented on api/src/opentrons/protocol_engine/commands/calibration/calibrate_module.py:
```diff
@@ -69,12 +69,21 @@ async def execute(self, params: CalibrateModuleParams) -> CalibrateModuleResult:
             labware_id=params.labwareId, well_name="B1"
         )
+        # Subtract the module offset included in get_well_position for new calibration
+        current_offset = self._state_view.modules.get_raw_module_offset(params.moduleId)
```
The problem with this is that it's doomed to break if we add more offsets somewhere in the future. What if we instead add a `get_nominal_well_position` and `get_nominal_module_position`, which don't apply calibration offsets?

vegano1 replied:
you're absolutely correct, good feedback

sfoster1 reviewed (changes requested):
The functionality looks good but I'm a little nervous doing this `get_raw_module_offset` and then subtracting it thing. It makes the code in the calibration command depend on the internal behavior of other modules. We're now depending on the implicit contract that the only offset from nominal in the well position is whatever `get_raw_module_offset` returns, which is a contract across two separate modules.

What if instead we have `get_nominal_well_position`, which probably internally depends on a `get_nominal_module_position` (and maybe raises `NotImplementedError` if you ask it for the nominal position of a well that's not something on a labware)? Now we've made the contract explicit: if you ask for `get_nominal_well_position()`, it _must_ take into account nothing but the predefined geometry.

I think that would be a little more robust to regressions.

vegano1 committed:
split up calibrated module offset from get_module_offset into its own function so we can get the module offset vector with or without calibrated module offsets
added get_labware_parent_position which has the same functionality as the previous get_labware_position but without the calibrated module offsets
get_labware_position now uses the new get_labware_parent_position method and adds the calibrated module offsets with the new get_labware_offset_vector function
this lets us use get_nominal_well_position to just get the uncalibrated well position when calibrating modules
updated the thermocycler calibration adapter definition to match the source dimensions

vegano1 committed:
split up calibrated module offset from get_module_offset into its own function so we can get the module offset vector with or without calibrated module offsets
added get_labware_parent_position which has the same functionality as the previous get_labware_position but without the calibrated module offsets
get_labware_position now uses the new get_labware_parent_position method and adds the calibrated module offsets with the new get_labware_offset_vector function
this lets us use get_nominal_well_position to just get the uncalibrated well position when calibrating modules
updated the thermocycler calibration adapter definition to match the source dimensions

vegano1 committed:
Merge branch 'RCORE-738_clear-offset-on-new-module-calibration' of https://github.com/Opentrons/opentrons into RCORE-738_clear-offset-on-new-module-calibration

vegano1 committed:
fix hardware-testing lint

sfoster1 commented on api/src/opentrons/protocol_engine/state/modules.py:
```diff
@@ -630,7 +630,16 @@ def get_dimensions(self, module_id: str) -> ModuleDimensions:
         return self.get_definition(module_id).dimensions
 
-    def get_module_offset(
+    def get_module_offset_vector(self, module_id: str) -> ModuleOffsetVector:
```
nit: can we call this `get_module_calibration_offset`? i'm getting confused by having both this and `get_module_offset`

SyntaxColoring replied:
Agreed. The only reason *`get_labware_offset_vector()`* is named like it is is because the official terminology is that Labware Position Check saves *labware offsets,* not *labware calibrations.* (🤷) But "calibration" is more specific than just "offset," so it's nice to say "calibration" when we can.

sfoster1 commented on api/src/opentrons/protocol_engine/state/geometry.py:
```diff
@@ -155,23 +168,35 @@ def get_well_position(
+        parent_pos = self.get_labware_parent_origin_position(labware_id)
+        origin_offset = self._labware.get_definition(labware_id).cornerOffsetFromSlot
+        well_def = self._labware.get_well_definition(labware_id, well_name)
+        return Point(
```
nit: could do `return parent_pos + origin_offset + well_def + Point(x=0, y=0, z=well_def.depth)` but it's up to you whether that's clearer or not

sfoster1 reviewed (approved):
This looks great to me aside from those couple nits! Nice work!

SyntaxColoring reviewed (approved):
Code makes sense to me. Thanks!

The overall approach also makes sense to me, but with a caveat: it only protects `loadModule` commands. That means, in the module calibration flows, we'll have to be careful that the app doesn't issue any other commands that would visit this module.

For example, if the app did a `moveToWell` command, that would still use the module's old calibration, right? The app *does* issue other commands like `moveToWell` during Labware Position Check. So we should make sure it doesn't need to do that for module calibration.

sanni-t commented on api/src/opentrons/protocol_engine/state/geometry.py:
```diff
@@ -90,11 +91,10 @@ def get_min_travel_z(
 
-    def get_labware_parent_position(self, labware_id: str) -> Point:
-        """Get the position of the labware's parent slot (deck or module)."""
+    def get_labware_parent_origin_position(self, labware_id: str) -> Point:
```
I would keep this `get_labware_parent_position` or make it `get_labware_parent_nominal_position`..

sanni-t commented on api/src/opentrons/protocol_engine/state/geometry.py:
```diff
@@ -108,20 +108,33 @@ def get_labware_parent_position(self, labware_id: str) -> Point:
+                slot_pos.z + module_offset.z,
             )
 
+    def get_labware_parent_position(self, labware_id: str) -> Point:
```
..and change this to `get_labware_parent_calibrated_position`

vegano1 replied:
kept it as is for consistency

sanni-t reviewed (approved):
LGTM! 
Just the two naming suggestions similar to @sfoster1 's

sanni-t commented:
> The overall approach also makes sense to me, but with a caveat: it only protects `loadModule` commands. That means, in the module calibration flows, we'll have to be careful that the app doesn't issue any other commands that would visit this module.
> 
> For example, if the app did a `moveToWell` command, that would still use the module's old calibration, right? The app _does_ issue other commands like `moveToWell` during Labware Position Check. So we should make sure it doesn't need to do that for module calibration.

Am I understanding your concern correctly that if a `moveToWell` command was issued immediately after a module calibration, in the same run, then the instrument might end up going to a position that was calculated with the calibration offsets that were loaded before the most recent module calibration (unless the module is reloaded or a new run is started after the calibration)? If so, then that's actually not what would happen. There is indeed a state update handler that would update the particular module's calibration offset in state after a successful module calibration so we'll start using the new offset right away for all moveTo commands.

vegano1 committed:
rename some functions

SyntaxColoring commented:
> Am I understanding your concern correctly that if a moveToWell command was issued immediately after a module calibration, in the same run, then the instrument might end up going to a position that was calculated with the calibration offsets that were loaded before the most recent module calibration (unless the module is reloaded or a new run is started after the calibration)? If so, then that's actually not what would happen. There is indeed a state update handler that would update the particular module's calibration offset in state after a successful module calibration so we'll start using the new offset right away for all moveTo commands.

No—my concern is that if a `moveToWell` command is issued *before* the `calibrateModule` command, the `moveToWell` command will use an old, possibly bad offset, while the `calibrateModule` command will use the nominal position. Ideally, I think they'd both use the nominal position.

For example, imagine if the app had a step where it did `moveToWell` and then said "OK user, I'm about to calibrate here." And then it...didn't calibrate there.
